File: tests/events_helper.py
from datetime import timedelta, datetime, timezone

import httpx

import config



def create_nearest_event_by_api(event_duration=timedelta(minutes=15), look_ahead_days=400) -> int:
    """Create a Kaltura event at the nearest available slot and return its ID."""
    ks = config.KALTURA_KS
    base_url = config.KALTURA_PUBLIC_API

    candidate, slot_end = get_nearest_free_slot(event_duration, look_ahead_days)

    response = httpx.post(
        f"{base_url}/events/create",
        headers={"Authorization": f"ks {ks}"},
        json={
            "name": "Test Event",
            "description": "Test description",
            "templateId": "tm1000",
            "startDate": candidate.isoformat(),
            "endDate": slot_end.isoformat(),
            "timezone": "Etc/GMT-0",
    },
        timeout=60,
    )
    assert response.status_code == 200, f"Invalid response on creating event: {response.text}"
    json = response.json()
    assert "event" in json and "id" in json["event"], f"Invalid response: {json}"
    return response.json()["event"]["id"]


def delete_event_by_api(event_id: int) -> None:
    httpx.post(
        f"{config.KALTURA_PUBLIC_API}/events/delete",
        headers={"Authorization": f"ks {config.KALTURA_KS}"},
        json={"id": event_id},
        timeout=60,
    ).raise_for_status()


def get_nearest_free_slot(
    event_duration: timedelta = timedelta(minutes=15),
    look_ahead_days: int = 400,
) -> tuple[datetime, datetime]:
    ks = config.KALTURA_KS
    base_url = config.KALTURA_PUBLIC_API

    now = datetime.now(tz=timezone.utc)
    window_end = now + timedelta(days=look_ahead_days)

    response = httpx.post(
        f"{base_url}/events/list",
        headers={"Authorization": f"ks {ks}"},
        json={
            "filter": {
                "startDateGreaterThanOrEqual": now.isoformat(),
                "endDateLessOrEqualThan": window_end.isoformat(),
            },
            "pager": {"limit": 15, "offset": 0},
            "orderBy": "+startDate",
        },
        timeout=10,
    )
    response.raise_for_status()
    events = response.json().get("events", [])

    booked: list[tuple[datetime, datetime]] = []
    for ev in events:
        start_raw = ev.get("startDate")
        end_raw = ev.get("endDate")
        if start_raw and end_raw:
            booked.append((
                datetime.fromisoformat(start_raw.replace("Z", "+00:00")),
                datetime.fromisoformat(end_raw.replace("Z", "+00:00")),
            ))

    candidate = now.replace(second=0, microsecond=0) + timedelta(minutes=1)
    while candidate < window_end:
        slot_end = candidate + event_duration
        if not any(s < slot_end and candidate < e for s, e in booked):
            return candidate, slot_end
        candidate += timedelta(minutes=1)

    raise RuntimeError(f"No free slot found in the next {look_ahead_days} days")

File: tests/functional/helpers/eventsHelper.ts
// Direct Kaltura REST helpers used to set up / tear down real events as test
// fixtures. These bypass the MCP server and talk to the Events API directly,
// mirroring the Python `events_helper.py`.

import { config } from '../config'

const MINUTE_MS = 60_000
const DEFAULT_DURATION_MS = 15 * MINUTE_MS
const DEFAULT_LOOK_AHEAD_DAYS = 400

export interface FreeSlot {
  start: Date
  end: Date
}

function authHeaders(): Record<string, string> {
  return {
    Authorization: `ks ${config.KALTURA_KS}`,
    'Content-Type': 'application/json',
  }
}

/** Format a Date as `YYYY-MM-DDTHH:MM:SSZ` (no milliseconds), matching the Python isoformat. */
function toKalturaIso(date: Date): string {
  return date.toISOString().replace(/\.\d{3}Z$/, 'Z')
}

async function postJson(path: string, body: unknown, timeoutMs: number): Promise<Response> {
  return fetch(`${config.KALTURA_PUBLIC_API}${path}`, {
    method: 'POST',
    headers: authHeaders(),
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(timeoutMs),
  })
}

/** Find the nearest free time slot not overlapping any existing event. */
export async function getNearestFreeSlot(
  durationMs: number = DEFAULT_DURATION_MS,
  lookAheadDays: number = DEFAULT_LOOK_AHEAD_DAYS,
): Promise<FreeSlot> {
  const now = new Date()
  const windowEnd = new Date(now.getTime() + lookAheadDays * 24 * 60 * MINUTE_MS)

  const response = await postJson(
    '/events/list',
    {
      filter: {
        startDateGreaterThanOrEqual: toKalturaIso(now),
        endDateLessOrEqualThan: toKalturaIso(windowEnd),
      },
      pager: { limit: 15, offset: 0 },
      orderBy: '+startDate',
    },
    10_000,
  )
  if (!response.ok) {
    throw new Error(`Failed to list events: ${response.status} ${await response.text()}`)
  }

  const listed = (await response.json()) as { events?: Array<{ startDate?: string; endDate?: string }> }
  const events = listed.events ?? []
  const booked: FreeSlot[] = events
    .filter((ev) => ev.startDate && ev.endDate)
    .map((ev) => ({ start: new Date(ev.startDate as string), end: new Date(ev.endDate as string) }))

  const candidate = new Date(now)
  candidate.setUTCSeconds(0, 0)
  candidate.setUTCMinutes(candidate.getUTCMinutes() + 1)

  while (candidate < windowEnd) {
    const slotEnd = new Date(candidate.getTime() + durationMs)
    const overlaps = booked.some((b) => b.start < slotEnd && candidate < b.end)
    if (!overlaps) {
      return { start: new Date(candidate), end: slotEnd }
    }
    candidate.setUTCMinutes(candidate.getUTCMinutes() + 1)
  }

  throw new Error(`No free slot found in the next ${lookAheadDays} days`)
}

/** Create a Kaltura event at the nearest available slot and return its ID. */
export async function createNearestEventByApi(
  durationMs: number = DEFAULT_DURATION_MS,
  lookAheadDays: number = DEFAULT_LOOK_AHEAD_DAYS,
): Promise<number> {
  const { start, end } = await getNearestFreeSlot(durationMs, lookAheadDays)

  const response = await postJson(
    '/events/create',
    {
      name: 'Test Event',
      description: 'Test description',
      templateId: 'tm1000',
      startDate: toKalturaIso(start),
      endDate: toKalturaIso(end),
      timezone: 'Etc/GMT-0',
    },
    60_000,
  )

  const text = await response.text()
  if (response.status !== 200) {
    throw new Error(`Invalid response on creating event: ${text}`)
  }
  const json = JSON.parse(text)
  if (!json.event?.id) {
    throw new Error(`Invalid response: ${text}`)
  }
  return json.event.id
}

export async function deleteEventByApi(eventId: number): Promise<void> {
  const response = await postJson('/events/delete', { id: eventId }, 60_000)
  if (!response.ok) {
    throw new Error(`Failed to delete event ${eventId}: ${response.status} ${await response.text()}`)
  }
}
